### PR TITLE
[NEW] [Apps-Engine] New Livechat event handlers

### PR DESCRIPTION
**Pull request by lolimay**
This is the changes of the PR https://github.com/RocketChat/Rocket.Chat.Apps-engine/pull/248 on the  RocketChat side. Check issue https://github.com/RocketChat/Rocket.Chat.Apps-engine/issues/234 for the details.

### CODE REVIEW

**Comment by lolimay** [comment]
The app for testing:
```ts
import {
    IAppAccessors,
    IHttp,
    ILogger,
    IPersistence,
    IRead,
} from '@rocket.chat/apps-engine/definition/accessors';
import { App } from '@rocket.chat/apps-engine/definition/App';
import {  ILivechatAssignAgentHandler, ILivechatContext, ILivechatRoomClosedHandler, ILivechatRoomStartedHandler, ILivechatUnassignAgentHandler, ILivechatRoom } from '@rocket.chat/apps-engine/definition/livechat';
import { IAppInfo } from '@rocket.chat/apps-engine/definition/metadata';

export class LivechatTestingApp extends App implements ILivechatRoomStartedHandler, ILivechatRoomClosedHandler, ILivechatAssignAgentHandler, ILivechatUnassignAgentHandler {
    constructor(info: IAppInfo, logger: ILogger, accessors: IAppAccessors) {
        super(info, logger, accessors);
    }

    public async executeLivechatAssignAgentHandler(data: ILivechatContext, read: IRead, http: IHttp, persistence: IPersistence) {
        console.log('executeLivechatAssignAgentHandler', { data });
    }

    public async executeLivechatUnassignAgentHandler(data: ILivechatContext, read: IRead, http: IHttp, persistence: IPersistence) {
        console.log('executeLivechatUnassignAgentHandler', { data });
    }

    public async executeLivechatRoomStartedHandler(data: ILivechatRoom, read: IRead, http: IHttp, persistence: IPersistence) {
        console.log('executeLivechatRoomStartedHandler', { data });
    }

    public async executeLivechatRoomClosedHandler(data: ILivechatRoom, read: IRead, http: IHttp, persistence: IPersistence) {
        console.log('executeLivechatRoomClosedHandler', { data });
    }
}
```

**Comment by lolimay** [comment]
Hey @renatobecker . I found that only the agent clicking the button `return` will call the `RoutingManager.unassignAgent` method ( thus triggering the `LivechatUnassignAgent` event). Is it by design? I think it should also trigger `LivechatUnassignAgent` event when someone `forward` the vistor to others. What do you think?

image

**Comment by renatobecker** [comment]
> Hey @renatobecker . I found that only the agent clicking the button `return` will call the `RoutingManager.unassignAgent` method ( thus triggering the `LivechatUnassignAgent` event). Is it by design? I think it should also trigger `LivechatUnassignAgent` event when someone `forward` the vistor to others. What do you think?
> 
> image

Hi @lolimay!

Yeah, I got your point, but we can't rely on the `RoutingManager.unassignAgent` only.
My suggestion is to use the [removeAgentFromSubscription](https://github.com/RocketChat/Rocket.Chat/blob/develop/app/livechat/server/lib/Helper.js#L151)  to trigger the `LivechatUnassignAgent` event because it is called in the `RoutingManager.unassignAgent`, as well on the `forwardRoomToAgent` and the `forwardRoomToDepartment`. What do you think? This is exactly what you're expecting for.

Thanks.

**Comment by lolimay** [comment]
Thanks @renatobecker ! It makes sense!

**Comment by lolimay** [comment]
Hello @renatobecker ! It works well now during my local testing. Can you please help test it too? so that we can solve possible potential issues before merging it.

File: app/livechat/server/lib/Livechat.js [hunk 1]
```diff
@@ -368,7 +368,7 @@ export const Livechat = {
 		Messages.createCommandWithRoomIdAndUser('promptTranscript', rid, closeData.closedBy);
 
 		Meteor.defer(() => {
-			Apps.getBridges().getListenerBridge().livechatEvent(AppInterface.ILivechatRoomClosedHandler, room);
+			Apps.getBridges().getListenerBridge().livechatEvent(AppInterface.IPostLivechatRoomClosed, room);
```
[the remark this answers is not in the document]
Comment (lolimay) [reply app/livechat/server/lib/Livechat.js:376]: Thank you!

**Pull request merged**